dxr/lines.py

```diff
@@ -35,6 +37,167 @@ def __repr__(self):
+    """Abstract superclass for a cross-reference attached to a run of text
+
+    Carries enough data to construct a context menu, highlight instances of
+    the same symbol, and show something informative on hover.
+
+    """
+    sort_order = 1
+    __slots__ = ['menu_data', 'hover', 'qualname_hash']
+    __metaclass__ = RefClassIdTagger
+
+    def __init__(self, tree, menu_data, hover=None, qualname=None, qualname_hash=None):
+        """
+        :arg menu_data: Arbitrary JSON-serializable data from which we can
+            construct a context menu
+        :arg hover: The contents of the <a> tag's title attribute. (The first
+            one wins.)
+        :arg qualname: A hashable unique identifier for the symbol surrounded
+            by this ref, for highlighting
+        :arg qualname_hash: The hashed version of ``qualname``, which you can
+            pass instead of ``qualname`` if you have access to the
+            already-hashed version
+
+        """
+        self.tree = tree
+        self.menu_data = menu_data
+        self.hover = hover
+        self.qualname_hash = hash(qualname) if qualname else qualname_hash
+
+    def es(self):
+        """Return a serialization of myself to store in elasticsearch."""
+        ret = {'plugin': self.plugin,
+               'id': self.id,
+               # Smash the data into a string, because it will have a
+               # different schema from subclass to subclass, and ES will freak
+               # out:
+               'menu_data': json.dumps(self.menu_data)}
+        if self.hover:
+            ret['hover'] = self.hover
+        if self.qualname_hash is not None:  # could be 0
+            ret['qualname_hash'] = self.qualname_hash
+        return ret
+
+    @staticmethod
+    def es_to_triple(es_data, tree):
+        """Convert ES-dwelling ref representation to a (start, end,
+        :class:`~dxr.lines.Ref` subclass) triple.
+
+        Return a subclass of Ref, chosen according to the ES data. Into its
+        attributes "menu_data", "hover" and "qualname_hash", copy the ES
+        properties of the same names, JSON-decoding "menu_data" first.
+
+        :arg es_data: An item from the array under the 'refs' key of an ES LINE
+            document
+        :arg tree: The :class:`~dxr.config.TreeConfig` representing the tree
+            from which the ``es_data`` was pulled
+
+        """
+        def ref_class(plugin, id):
+            """Return the subclass of Ref identified by a combination of
+            plugin and class ID."""
+            plugins = all_plugins()
+            try:
+                return plugins[plugin].refs[id]
+            except KeyError:
+                warn('Ref subclass from plugin %s with ID %s was referenced '
+                     'in the index but not found in the current '
+                     'implementation. Ignored.' % (plugin, id))
+
+        payload = es_data['payload']
+        cls = ref_class(payload['plugin'], payload['id'])
```

Comment:
Now that we have fewer units in play that the index cares about (just one Ref subclass rather than a class for almost every menu item), I think we can just have this fail hard when it fails. When you mess with a Ref subclass in a non-backward-compatible way, bump the format version.